### DISCUSSION TITLE
Particle size distribution and experiment

### DISCUSSION THREAD

**Discussion by CaXYZ** [pybamm-team/PyBaMM#2577] [Q&A] [answered]
Hi all,

I would like to evaluate the DFN-Full cell model with particle size distribution via  a PyBaMM experiment.  I used the notebook provided for [DFN with particle size distribution](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/models/DFN-with-particle-size-distributions.ipynb) as a reference and added a PyBaMM experiment:

````
# import

import pybamm
import numpy as np
import matplotlib.pyplot as plt



# choose option(s)
options = {
    "particle size": "distribution"
}

# load model
model = pybamm.lithium_ion.DFN(options=options, name="MP-DFN")



# Base parameter set (no distribution parameters by default)
params = pybamm.ParameterValues("Marquis2019")

# Add distribution parameters to the set, with default values (lognormals)
params = pybamm.get_size_distribution_parameters(params)


# Define experiment
experiment=pybamm.Experiment(
    [('Rest for 10 s')
     ]+    
    [
        ("Charge at C/3 for 0.1 hours or until 4.3 V",),
    ]*1
)

# load parameter values into simuluation
solver = pybamm.CasadiSolver(mode="fast")
sim = pybamm.Simulation(model, experiment=experiment, parameter_values=params, solver=solver)

# solve
sim.solve(t_eval=[0, 3500])


# plot some variables that depend on particle size
output_variables = [
    "Negative particle surface concentration distribution",
    "Positive particle surface concentration distribution",
    "X-averaged negative particle surface concentration distribution",
    "Negative area-weighted particle-size distribution",
    "Positive area-weighted particle-size distribution",
    "Terminal voltage [V]",
]

sim.plot(output_variables=output_variables)
````

Without the experiment, the simulation runs perfectly fine. However, when I add the experiment, the following error message is printed: 
`NotImplementedError: Shape not recognized for y[1:12001] (note processing of 3D variables is not yet implemented)`

Am I doing something wrong, or can one currently not use PyBaMM experiment with PSD?

Best regards
Carsten

**Comment by rtimms** [comment, marked as answer]
See #1549

**Reply by CaXYZ** [reply]
Thank you for the input.

**Comment by Patrick-Moore1** [comment]
I found a work around that is easier than changing PyBaMM to accept 3D variables as suggested in #1549. I also encountered the same issue when trying to run a multistep experiment, and the best solution that I found is to directly create the current vs. time function that you expect (or know from experiment) using the pybamm.Interpolant function and then feed this into the simulation. This allows you to change the current without PyBaMM having to "pass" conditions between experimental steps which circumvents the annoying functions that give the 3D variables errors.

**Reply by CaXYZ** [reply]
Thank you for this idea :)
